graph/pull.go

```diff
@@ -27,10 +18,28 @@ type ImagePullConfig struct {
 	OutStream   io.Writer
 }
 
+type Puller interface {
+	// Pull tries to pull `canonicalImageName` with tag `tag`, from `endpoint` with the config specified by `imagePullConfig`.
+	// `name` is the image's name on the server-side
+	// `sf` is an object that allows to stream back to the client (to be removed)
+	// Pull returns an error if any, as well as a boolean that determines whether to retry Pull on the next configured endpoint.
+	Pull(canonicalImageName, name, tag string, endpoint registry.APIEndpoint, imagePullConfig *ImagePullConfig, sf *streamformatter.StreamFormatter) (fallback bool, err error)
```

Comment:
@tiborvass A few comments about this interface:

1. What is the difference between `canonicalImageName` and `name`? Should `name` be called `remoteName`? Since we've resolved the endpoint, do we need both? Or, could we always calculate  `canonicalImageName` from `name?
2. Can we pull `imagePullConfig` and `sf` up to the `Puller` creation method? They seem to be an argument to every method and really belong on the struct.
3. It looks like `endpoint` is passed to `NewPuller`, as well as `repoInfo`. Can they change or do they also remain static on the puller. If `repoInfo` is static, why even pass in an image name? (unless i misunderstand the role of `repoInfo`).